transport_plugins/jlink/iotile_transport_jlink/jlink_background.py

```diff
@@ -262,76 +188,142 @@ def _debug_read_memory(self, device_info, _control_info, args, _progress_callbac
+    async def debug_write_memory(self, _device_info, _control_info, args):
         memory_region = args.get('region').lower()
 
         if memory_region == 'external' or memory_region == 'flash':
             raise ArgumentError("Writing to flash/external is not currently supported.")
-        self._jlink.memory_write(args.get('address'), args.get('data'))
+        await self._loop.run_in_executor(
+            self._jlink.memory_write, args.get('address'), args.get('data'))
 
-    def _program_flash(self, _device_info, _control_info, args, progress_callback):
+    async def program_flash(self, _device_info, _control_info, args):
         base_address = args.get('base_address')
         data = args.get('data')
 
         if base_address is None or data is None:
             raise ArgumentError("Invalid arguments to program flash, expected a dict with base_address and data members")
 
+        conn_string = self._jlink_adapter._get_property(self._jlink_adapter._connection_id, 'connection_string')
+
         def _internal_progress(action, _prog_string, percent):
             """Convert jlink progress callback into our own (finished, total) format."""
-            action_table = {'Compare': 1, 'Erase': 2,
-                            'Flash': 3, 'Verify': 4}
+            action_table = {'Compare': 0, 'Erase': 1,
+                            'Flash': 2, 'Verify': 3}
 
-            if progress_callback is not None:
-                progress_callback(action_table.get(action, 4)*percent, 400)
+            event = {
+                "operation": "debug",
+                "finished": action_table.get(action, 3) * 100 + percent,
+                "total": 400
+            }
+            self._jlink_adapter.notify_event(conn_string, 'progress', event)
 
         logger.info("Flashing %d bytes starting at address 0x%08X", len(data), base_address)
-        self._jlink.flash(data, base_address, _internal_progress)
-        self._jlink.reset(halt=False)
+        await self._loop.run_in_executor(
+            self._jlink.flash, data, base_address, _internal_progress)
+        await self.reset(halt=False)
+
+    @staticmethod
+    def _connect_jlink_blocking(jlink_serial, jlink_name):
+        jlink = pylink.JLink()
+        jlink.open(jlink_serial)
+        jlink.set_tif(pylink.enums.JLinkInterfaces.SWD)
+        jlink.connect(jlink_name)
+        jlink.set_little_endian()
+
+        return jlink
+
+    async def connect_jlink(self, jlink_serial, jlink_name):
+        self._jlink = await self._loop.run_in_executor(
+            self._connect_jlink_blocking, jlink_serial, jlink_name)
+
+        return self._jlink
 
-    def _read_memory(self, start_address, length, chunk_size=4, join=True):
+    async def close_jlink(self):
+        if self._jlink is not None:
+            await self._loop.run_in_executor(self._jlink.close)
+
+    async def start_polling(self, control_info, step_timeout=0.1):
```

Comment:
I see here the counting approach to handing polling.  That works and you can ignore the previous comment I made about needing to do that, I hadn't read this far in the code yet.  I still think you need to individually track the status of the various interfaces but the overall polling approach makes sense to me now.